src/core/eme/create_session.ts

```diff
@@ -0,0 +1,213 @@
+  };
+}
+
+export type IGetSessionEvent =
+  ICreatedSessionEvent |
+  IPersistentSessionRecoveryEvent;
+
+/**
+ * If session creating fails, retry once session creation/loading.
+ * Emit true, if it has succeeded to load, false if there is no data for the
+ * given sessionId.
+ * @param {string} sessionId
+ * @param {MediaKeySession} session
+ * @returns {Observable}
+ */
+function loadPersistentSession(
+  sessionId: string,
+  session: MediaKeySession|IMediaKeySession
+) : Observable<boolean> {
+  return Observable.defer(() => {
+    log.debug("eme: load persisted session", sessionId);
+    return castToObservable(session.load(sessionId));
+  });
+}
+
+/**
+ * Create a new Session on the given MediaKeys, corresponding to the given
+ * initializationData.
+ * If session creating fails, remove the oldest MediaKeySession loaded and
+ * retry.
+ *
+ * /!\ This only creates new sessions.
+ * It will fail if $loadedSessions already has a MediaKeySession with
+ * the given initializationData.
+ * @param {Uint8Array} initData
+ * @param {string} initDataType
+ * @param {Object} mediaKeysInfos
+ * @returns {Observable}
+ */
+function createSession(
+  initData: Uint8Array,
+  initDataType: string,
+  mediaKeysInfos: IMediaKeysInfos
+) : Observable<IGetSessionEvent> {
+  const {
+    keySystemOptions,
+    keySystemAccess,
+    mediaKeys,
+    sessionStorage,
+  } = mediaKeysInfos;
+
+  if (mediaKeys.createSession == null) {
+    throw new Error("Invalid MediaKeys implementation: Missing createSession");
+  }
+
+  const mksConfig = keySystemAccess.getConfiguration();
+  const sessionTypes = mksConfig.sessionTypes;
+  const hasPersistence = (
+    sessionTypes && arrayIncludes(sessionTypes, "persistent-license")
+  );
+  const sessionType = hasPersistence && keySystemOptions.persistentLicense ?
+    "persistent-license" : "temporary";
+
+  log.debug(`eme: create a new ${sessionType} session`);
+
+  const session = (mediaKeys as any /* TS bug */).createSession(sessionType);
+  $loadedSessions.add(initData, initDataType, session);
+
+  if (!hasPersistence || !sessionStorage || !keySystemOptions.persistentLicense) {
+    return Observable.of({
+      type: "created-session" as "created-session",
+      value: { mediaKeySession: session },
+    });
+  }
+
+  const storedEntry = sessionStorage.get(initData, initDataType);
+  if (!storedEntry) {
+    return Observable.of({
+      type: "created-session" as "created-session",
+      value: { mediaKeySession: session },
+    });
+  }
+
+  return loadPersistentSession(storedEntry.sessionId, session)
+    .catch((_) => {
+      $loadedSessions.closeSession(session);
+      sessionStorage.delete(initData, initDataType);
+
+      const newSession =
+        (mediaKeys as any /* TS bug */).createSession("persistent-license");
+      $loadedSessions.add(initData, initDataType, newSession);
+
+      return Observable.of({
+        type: "created-session" as "created-session",
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this unnecessary cast. ([typescript:S4325](https://sonarcloud.io/organizations/rx-player/rules#rule_key=typescript%3AS4325))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=rx-player&pullRequest=187&issues=AWMrRtWamtQEh111nXqA&open=AWMrRtWamtQEh111nXqA)

[ISSUE_KEY]: AWMrRtWamtQEh111nXqA